Half expr casting

delcypher opened the pull request:
This 

* Adds a rounding mode argument to the ``Cast`` node
* Adds a software implementation of casting between ``float``/``double`` to ``float16``
* Adds support for using x86 vector intrinsics to do the above conversion when possible
* Adds a whole load of tests!

@abadams @abestephensg @zvookin

abadams commented on src/CodeGen_LLVM.cpp:
```diff
@@ -1195,6 +1204,94 @@ void CodeGen_LLVM::visit(const Cast *op) {
+            // doesn't support this natively LLVM will emit a call to
+            // ``__gnu_f2h_ieee`` which isn't available as we don't user
+            // compiler-rt so we use our own implementation
+            user_assert(src.width == 1 && src.width == dst.width) <<
```
Use the scalarize method to handle this case.

delcypher replied:
I didn't know about that method. That seems a better was of handling this situation rather than bailing out with a ``user_assert``.

abadams commented on src/CodeGen_LLVM.cpp:
```diff
@@ -1195,6 +1204,94 @@ void CodeGen_LLVM::visit(const Cast *op) {
+            user_assert(src.width == 1 && src.width == dst.width) <<
+                "The target requires a software implementation of float16_t conversion"
+                " but vectorization has been requested which is not possible\n";
+            llvm::Function* convFunc = nullptr;
```
Please use underscores instead of camelCase for local variables and functions throughout this PR.

delcypher replied:
Is it fine to do this as an additional commit? I don't really want to rebase and redo all the commits.

abadams replied:
Yes, that's fine.

abadams commented on src/CodeGen_ARM.cpp:
```diff
@@ -1261,4 +1261,14 @@ int CodeGen_ARM::native_vector_bits() const {
     return 128;
 }
 
+bool CodeGen_ARM::target_needs_software_cast_from_float16_to(Type t) const {
```
Are we ever going to encounter a target with hardware support for casts from float16 but not hardware support for casts to float16?

I suggest simplifying this pair of methods into something like "has_hardware_float16_cast"

delcypher replied:
> Are we ever going to encounter a target with hardware support for casts from float16 but not hardware support for casts to float16?

Yes, sort of... There are two methods because of 

- how this PR handles the conversion behavior on x86. It is **not symmetrical**.
-  When downcasting **we need to know the rounding mode**, but it's irrelevant we upcasting.

Take a look at the contents of ``CodeGen_X86::target_needs_software_cast_from_float16_to(Type t)`` and ``CodeGen_X86::target_needs_software_cast_to_float16_from(Type t, RoundingMode rm)`` to understand this.

Briefly though, the ``double -> float16`` and ``float16 -> double`` conversions are **not** symmetrical in my implementation on x86.

* ``float16 -> double`` we can do in hardware because we can use ``vcvtph2ps`` (x86-64 fp16c intrinsic) to get a single precision and then convert (``fpext`` LLVM IR instruction) to double. Upcasting twice is legal because there is no loss in precision.

* ``double -> float16`` cannot be done in hardware because ``vcvtph2ps`` only works with ``float`` and rounding twice (``double`` -> ``float`` -> ``float16``) is not acceptable because that would not respect the rounding mode. Therefore the software implementation is **always used here**.

* ``float -> float16`` can be done in hardware provided the rounding mode is not "round to nearest, ties to away". All other rounding modes are supported by ``vcvtps2ph`` but the other rounding mode must be handled by the software implementation.

* ``float16 -> float`` can be done in hardware using the ``vcvtps2ph`` intrinsic.

abadams commented on src/CodeGen_X86.cpp:
```diff
@@ -306,10 +306,249 @@ void CodeGen_X86::visit(const Select *op) {
+            // Cast argument to <8 x i16> as expected by the intrinsic
+            valueToCast = builder->CreateBitCast(valueToCast, i16x8);
+
+            // Can't use call_intrin() because it asserts the number of elements
```
I think there's a much simpler implementation to be had here. Codegen for vectorized call nodes already pads things out to native vectors, or slices things up into native vectors, or scalarizes. It does this according to what functions are found to exist in the runtime. E.g. if only a scalar function exists, it uses that on each lane. If a vector function exists, it pads out the vectors as needed and calls that. See the CodeGen_LLVM.cpp around line 2708

So what you could do is just always codegen a Call node of halide_float16_bits_to_float, and then in runtime/x86_f16c.ll you provide a function halide_float16_bits_to_float which has the right ll for the hardware conversion, and functions with suffixes like halide_float16_bits_to_floatx8, etc which do the vectorized versions. You'd need to rejigger get_initial_module_for_target to include x86_f16c.ll instead of float16_t.cpp for x86 with f16c.

This way new platforms that support float16 in hardware only need to provide implementations of halide_float16_bits_to_float and the like in their runtime modules, and set up get_initial_module_for_target to use those runtime modules, and any vector versions will get used as appropriate automatically.

For an example of how this works, see that floor_f32x4 exists in x86_sse41.ll, and floor_f32x8 exists in x86_avx.ll. codegen of a Call node to floor_f32 of some arbitrary vector width uses whichever ones exist intelligently.

delcypher replied:
> For an example of how this works, see that floor_f32x4 exists in x86_sse41.ll, and floor_f32x8 exists in x86_avx.ll. codegen of a Call node to floor_f32 of some arbitrary vector width uses whichever ones exist intelligently.

Where is the logic for this "intelligent decision"?

``floor_f32(...)`` in  just calls ``@llvm.floor.f32(...)``.  I don't see how ``floor_f32x4`` is going to get called if a call to ``floor_f32()`` is emitted...

delcypher replied:
@abadams Sorry I see where this is now. This approach would work for the upcast but I'm not really sure this would work for doing the downcast because the vectorized instruction doesn't work with all rounding modes. The vectorized down cast could be made to potentially branch on the rounding mode argument and call the software implementation but this moves the decision on what to call to the runtime rather than at codegen time (unless LLVM is able propagate the constant argument to intrinsic, do dead code elimination and then inline).

delcypher replied:
@abadams It just occurred to me there is an additional issue here. On other platforms there may be a native instruction for conversion between ``float16`` and ``single``/``double`` but that might not be a vector instruction. An example is the ``FCVT`` instruction on AArch64. That means you need a way of picking the non vector hardware instruction over the software implementation and what you've proposed doesn't offer this IIUC.

abadams commented:
I had a longer comment that seems to have gotten lost by me tabbing around before hitting the button. Here's the concise version:

I'm always very conservative about adding fields to core IR nodes. For the rounding mode flag, I'm concerned with consistency between float->float16 an float->int casts. I think there are two reasonable paths:

1) Rejigger float -> int rounding to use the rounding mode field instead of Call nodes to round, trunc, ceil, floor. Those functions would still exist in the front-end, but they'd just make cast nodes with the appropriate flag.

2) Don't add the field to cast nodes, and have float -> float16 rounding always use some reasonable default rounding mode. Add a math library function that gets you the others (i.e. give the front-end a way to make calls to float16_bits_from_float).

I think 1 is probably better.

delcypher commented:
> I'm always very conservative about adding fields to core IR nodes. For the rounding mode flag, I'm concerned with consistency between float->float16 an float->int casts. I think there are two reasonable paths:

> 1) Rejigger float -> int rounding to use the rounding mode field instead of Call nodes to round, trunc, ceil, floor. Those functions would still exist in the front-end, but they'd just make cast nodes with the appropriate flag.

I think I noted in one of my commits that this attribute is only respected when converting from ``float``/``double`` to ``float16``. But that behavior probably isn't desirable and using the attribute for ``float -> int`` casts sounds reasonable. We need to define what the rounding modes mean though as they are meant for conversion where the result is a float, perhaps...

* Round up - Same as ``ceil()``.
* Round down - Same as ``floor()``.
* Round to zero - if input is positive acts like ``floor()`` otherwise acts like ``ceil()``.
* Round to nearest, ties to even - Round to nearest integer, ties pick result zero least significant bit.
* Round to nearest, ties to away - Round to nearest integer, ties away from zero.

What are the semantics when the input is ``NaN``, ``+Inf``, ``-Inf``?

delcypher commented:
@abadams How would you like to proceed with this? There still implementation issues (i.e. your suggestion for simpler implementation in it's current form isn't general enough).

One thing I could do is 

1. Close this PR and extract the non-controversial software runtime functions that do downcasting of  float/double to half runtime functions and get that merged
2. Redo the support for rounding mode in the Cast IR node that addresses your concerns and get that merged.
3. Once a decision is made on how hardware support for half conversion should be implemented (you obviously aren't happy with my current implementation) I can do that.

abadams commented:
Sorry, I've been unresponsive on this because I'm in Australia on paternity
leave and mostly away from a computer. I'll be back next Wednesday.
On Sep 20, 2015 6:54 PM, "Dan Liew" <notifications@github.com> wrote:

> @abadams <https://github.com/abadams> How would you like to proceed with
> this? There still implementation issues (i.e. your suggestion for simpler
> implementation in it's current form isn't general enough).
>
> One thing I could do is
>
>    1. Close this PR and extract the non-controversial software runtime
>    functions that do downcasting of float/double to half runtime functions and
>    get that merged
>    2. Redo the support for rounding mode in the Cast IR node that
>    addresses your concerns and get that merged.
>    3. Once a decision is made on how hardware support for half conversion
>    should be implemented (you obviously aren't happy with my current
>    implementation) I can do that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/932#issuecomment-141759191>.
>

abadams commented:
Back from Australia and digging through my todo list. 

I've hard a hard time deciding how important rounding modes are, so I had a talk today with one of the creators of OpenEXR, which uses its own software float16 and has existed for a long time now. In fact it won a technical oscar in 2007. He says that they implemented float -> float16 as round-to-nearest, intending to add other rounding modes when people asked for them, and nobody ever did. He's of the opinion that round-to-nearest is enough and supporting more rounding modes is unnecessary complexity. He expects float->float16 rounding to act like double->float rounding, where nobody ever bothers to specify a rounding mode. 

Dillon also says he used float16 in cuda heavily in a previous job and only ever used round-to-nearest.

In Halide, I don't think you'd ever want to use a rounding mode that goes through a software path when a hardware path is available. An unexpected performance cliff is worse than pixels coming out slightly different. This is in line with our -ffast-math philosophy and other floating point approximations we make.

So I think the right thing is:
- a runtime function declaration in HalideRuntime.h that converts from float to float16 with some flavor of round to nearest (probably ties to even), so there'd be no rounding mode argument or enum.
- multiple implementations of this runtime function. x86_f16c.ll will define a scalar version that uses the hardware instruction, and also a vectorized version (suffixed with x8, or whatever the width is). 
- The existing software implementation will exist as another runtime module
- Other backends are free to define their own runtime module to use their own hardware version (e.g. ptx_dev.ll)
- During initial module construction, (LLVM_Runtime_Linker), we'll either select the software implementation or one of the hardware implementations depending on the Target and its feature flags. Then codegen just converts the cast to an Extern call (e.g. like floor, or round), and the call_intrin magic automatically deals with vectors of arbitrary width intelligently, so the amount of new code in CodeGen_LLVM is minimal (just convert that type of cast into a call node instead), and there'd be no new code in CodeGen_X86.
- codegen does casts from doubles to float16 via float (we care more about speed than precision)

I understand that your internship is over, and this is quite different to what you've done, so it's unreasonable to expect that you'd be interested in doing this. Happy to do it myself if so - the hard part of the code is the same (the software implementation and tests).

delcypher commented:
@abadams 

> He expects float->float16 rounding to act like double->float rounding, where nobody ever bothers to specify a rounding mode.

The fact that most people don't specify rounding modes is probably a combination of several factors

* Most people don't know there's an option to control this.
* In C it's not exposed as a per cast feature but instead via functions that control the thread local state (``fesetround()`` and ``fegetround()``), which IMHO is not a very nice way to do it.
* "Round to nearest, ties to even" (RNE) is a very sensible default.
* On some platforms to RNE is the only rounding mode that's supported.

Just because C doesn't do it nicely doesn't necessarily mean that Halide also has to do this. I think it would be quite for the ``Cast`` node to take a rounding mode which could be used for other conversion (e.g. ``float`` -> ``int``).

However I understand that having extra code in Halide that is usually not needed could turn into an unnecessary maintenance burden. So I won't fight for this.

> I understand that your internship is over, and this is quite different to what you've done, so it's unreasonable to expect that you'd be interested in doing this. Happy to do it myself if so - the hard part of the code is the same (the software implementation and tests).

I'd quite like the software implementation to be committed by me (quite difficult to get right). I could extract the runtime software implementation from this PR for convert ``float`` and ``double`` to half into a separate PR. I'd keep a commit in there that has the implementation of all rounding modes and then I can make a separate commit that removes them and just uses RNE. That way you could potentially revert the commit to get support for all the rounding modes.

The rest (target specific implementations) I don't really have time to do so I'm happy to leave that with you. You can probably use this PR as a rough guide for what generated vectorized code for x86_64 (with fp16c) should look like.

How does that sound?

abadams commented:
That sounds good.

On Mon, Sep 28, 2015 at 6:11 AM, Dan Liew <notifications@github.com> wrote:

> @abadams <https://github.com/abadams>
>
> He expects float->float16 rounding to act like double->float rounding,
> where nobody ever bothers to specify a rounding mode.
>
> The fact that most people don't specify rounding modes is probably a
> combination of several factors
>
>    - Most people don't know there's an option to control this.
>    - In C it's not exposed as a per cast feature but instead via
>    functions that control the thread local state (fesetround() and
>    fegetround()), which IMHO is not a very nice way to do it.
>    - "Round to nearest, ties to even" (RNE) is a very sensible default.
>    - On some platforms to RNE is the only rounding mode that's supported.
>
> Just because C doesn't do it nicely doesn't necessarily mean that Halide
> also has to do this. I think it would be quite for the Cast node to take
> a rounding mode which could be used for other conversion (e.g. float ->
> int).
>
> However I understand that having extra code in Halide that is usually not
> needed could turn into an unnecessary maintenance burden. So I won't fight
> for this.
>
> I understand that your internship is over, and this is quite different to
> what you've done, so it's unreasonable to expect that you'd be interested
> in doing this. Happy to do it myself if so - the hard part of the code is
> the same (the software implementation and tests).
>
> I'd quite like the software implementation to be committed by me (quite
> difficult to get right). I could extract the runtime software
> implementation from this PR for convert float and double to half into a
> separate PR. I'd keep a commit in there that has the implementation of all
> rounding modes and then I can make a separate commit that removes them and
> just uses RNE. That way you could potentially revert the commit to get
> support for all the rounding modes.
>
> The rest (target specific implementations) I don't really have time to do
> so I'm happy to leave that with you. You can probably use this PR as a
> rough guide for what generated vectorized code for x86_64 (with fp16c)
> should look like.
>
> How does that sound?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/932#issuecomment-143739336>.
>

abadams commented:
Superceded by #970
